server/src/output_stream.rs

```diff
@@ -0,0 +1,112 @@
+use std::sync::Arc;
+
+use tokio::sync::mpsc::Sender;
+use tokio::sync::{Notify, RwLock};
+
+use common::output_event::Stream as OutputStream;
+use common::OutputEvent;
+
+/// Internal state of the `OutputHandler`
+struct State {
+    history: Vec<Vec<u8>>,
+    completed: bool,
+}
+
+/// Handles a single output stream
+pub struct OutputHandler {
+    /// Stdout or Stderr
+    stream_type: OutputStream,
+    /// Internal state
+    state: RwLock<State>,
+    /// State change notification
+    notify: Notify,
+}
+impl OutputHandler {
+    pub fn new(stream_type: OutputStream) -> Self {
+        Self {
+            stream_type,
+            state: RwLock::new(State {
+                history: Vec::new(),
+                completed: false,
+            }),
+            notify: Notify::new(),
+        }
+    }
+
+    /// Sets up a new OutputHandler and the stream into it from any `AsyncRead`-object,
+    /// usually either ChildStdout or ChildStderr.
+    pub fn setup<R>(stream_type: OutputStream, pipe: R) -> Arc<Self>
+    where
+        R: tokio::io::AsyncRead + Unpin + Send + 'static,
+    {
+        use futures_util::StreamExt;
+
+        let arc_self = Arc::new(OutputHandler::new(stream_type));
+
+        let inner = arc_self.clone();
+        tokio::spawn(async move {
+            let mut out = tokio_util::io::ReaderStream::new(pipe);
+            while let Some(value) = out.next().await {
+                let x = value.expect("Process output error");
+                inner.push(x.to_vec()).await;
+            }
+            inner.complete().await;
+        });
+
+        arc_self
+    }
+
+    /// Push new data to the history, notifying all waiting processes
+    pub async fn push(&self, data: Vec<u8>) {
```

Comment:
Not sure push/complete should be public functions as they look internal only, especially since the process can panic if miss-used.

Reply:
They are not necessarily internal only; if you would like to use `OutputHandler` for something that doesn't implement `AsyncRead` (for instance, stdlib process output), then those would be still useful. The panic conditions are documented in the doc comment. Although they are used here only internally, there is nothing from with them being public (from OOP perspective, anyway). On the other hand, it might be a bit cleaner to keep them private until they are used outside the class, since the whole `server` is a binary crate, and doesn't provide an code-level API anyway.